bookdown/05-extending/05-extending-mlr3pipelines.Rmd

```diff
@@ -1,5 +1,470 @@
+    }
+  )
+)
+```
+
+_(Note for the observant: If you check `PipeOpScale.R` from the `mlr3pipelines` package, you will notice that is uses "`get("type")`" and "`get("id")`" instead of "`type`" and "`id`", because the static code checker on CRAN would otherwise complain about references to undefined variables. This is a "problem" with `data.table` and not exclusive to `mlr3pipelines`.)_
+
+We can, again, create a new `Graph` that uses this `PipeOp` to test it.
+Compare the resulting data to the original "iris" `Task` data printed at the beginning:
+
+```{r 05-extending-025}
+gr = Graph$new()
+gr$add_pipeop(PipeOpScaleAlways$new())
+
+result = gr$train(task)
+
+result[[1]]$data()
+```
+
+### Special Case: Preprocessing with Simple Train
+
+It is possible to make even further simplifications for many `PipeOp`s that perform mostly the same operation during training and prediction.
+The point of `Task` preprocessing is often to modify the training data in mostly the same way as prediction data (but in a way that *may* depend on training data).
+
+Consider constant feature removal, for example: The goal is to remove features that have no variance, or only a single factor level.
+However, what features get removed must be decided during *training*, and may only depend on training data.
+Furthermore, the actual process of removing features is the same during training and prediction.
+
+A simplification to make is therefore to have a function `get_state(task)` which sets the `$state` slot during training, and a `transform(task)` function, which gets called both during training *and* prediction.
+This is done in the `PipeOpTaskPreprocSimple` class.
+Just like `PipeOpTaskPreproc`, one can inherit from this and overload these functions to get a `PipeOp` that performs preprocessing with very little boilerplate code.
+
+Just like `PipeOpTaskPreproc`, `PipeOpTaskPreprocSimple` offers the possibility to instead overload the `get_state_dt(dt, levels)` and `transform_dt(dt, levels)` functions (and optionally, again, the `select_cols(task)` function) to operate on `data.table` feature data instead of the whole `Task`.
+
+Even some methods that do not use `PipeOpTaskPreprocSimple` *could* work in a similar way: The `PipeOpScaleAlways` example from above will be shown to also work with this paradigm.
+
+#### Example: `PipeOpDropConst`
+
+A typical example of a preprocessing operation that does almost the same operation during training and prediction is an operation that drops features depending on a criterion that is evaluated during training.
+One simple example of this is dropping constant features. Because the `mlr3` `Task` class offers a flexible view on underlying data, it is most efficient to drop columns from the task directly using its `$select()` function, so the `get_state_dt(dt, levels)` / `transform_dt(dt, levels)` functions will *not* get used; instead we overload the `get_state(task)` and `transform(task)` functions.
```

Comment:
```suggestion
One simple example of this is dropping constant features. 
Because the `mlr3` `Task` class offers a flexible view on underlying data, it is most efficient to drop columns from the task directly using its `$select()` function, so the `get_state_dt(dt, levels)` / `transform_dt(dt, levels)` functions will *not* get used; instead we overload the `get_state(task)` and `transform(task)` functions.
```